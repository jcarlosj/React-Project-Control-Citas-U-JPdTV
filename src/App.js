import React, { Fragment, useState, useEffect } from 'react';

/** Components */
import AppointmentForm from './components/AppointmentForm';
import Appointment from './components/Appointment';

function App() {

    let appointmentsLocalStorage = JSON .parse( localStorage .getItem( 'appoinments' ) );   // Retorna null si no hay una cadena registrada
    /** Valida si NO hay citas en el LocalStorage */
    if( ! appointmentsLocalStorage ) {
        appointmentsLocalStorage = [];      // Inicializa Array Vacio
    }

    /** State: Citas Pendientes */
    const [ statusPendingAppointments, setPendingAppointments ] = useState( appointmentsLocalStorage );

    /** useEffect() se ejecuta y realiza algunas operaciones
     *   - Cuando el Componente a Cargado completamente (como el window.onload de JavaScript o el document.ready de jQuery).
     *   - Cuando algo cambia en el componente.
     *   - Si deseamos que se ejecute una sola vez debemos pasar como segundo parámetro un Array vacío, de lo contrario se ejecutará ciclicamente.
     *   - En dicho Array se van a colocar las dependencias que le hará seguimiento y que cambiarán para que el useEffect se ejecute.
     * 
     * NOTA:* Si estás familiarizado con el ciclo de vida de las clases de React y sus métodos, el Hook useEffect equivale a componentDidMount, componentDidUpdate y componentWillUnmount combinados.
    */
    useEffect( () => {                      // Siempre se define con un CallBack de Arrow Function
        console .log( 'Listo' );     

        /** Valida SI hay citas en el LocalStorage */
        if( appointmentsLocalStorage ) {
            localStorage .setItem( 'appointments', JSON .stringify( statusPendingAppointments ) );
        }
        else {
            localStorage .setItem( 'appointments', JSON .stringify( [] ) );
        }
        
    }, [ statusPendingAppointments ] );     // Array de dependencias que cambian

    /** Agrega solicitud a citas pendientes */
    const addAppointment = ( appointment ) => {
        console .log( 'App', appointment );

        /** Establece cambios en el State es decir. Asigna Cita solicitada a listado de Citas Pendientes */
        setPendingAppointments([
            ...statusPendingAppointments,   // Spread
            appointment                     // Enviado por el Componente Hijo
        ]);
    }

    /** Elimina citas atendidas por su ID */
    const deleteAppointmentAttended = ( id ) => {
        console .log( 'Atendida cita con ID', id );

        /** Obtiene todas las citas que faltan por atender, excluyendo la cita atendida (eliminada) */
        const pendingAppointments = statusPendingAppointments .filter( appointment => appointment .id !== id );

        /** Establece cambios en el State es decir. Asigna Citas Pendientes */
        setPendingAppointments( pendingAppointments );

    }

    /**  */
    const title = statusPendingAppointments .length === 0 
        ?   'No hay citas pendientes'
        :   'Listado citas';

  return (
    <Fragment>
        <h1>Administrar citas</h1>
        <div className="container">
            <div className="row">
                <div className="one-half column">
                    <AppointmentForm 
                        addAppointment={ addAppointment }       // Usamos Props para pasar la funcion al componente y establecer una comunicación con el padre 'App'
                    />
                </div>
                <div className="one-half column">
                    <h3>{ title }</h3>
                    {   statusPendingAppointments .map( appointment => (        // Return Implicito usando Parentesis
                        <Appointment 
                            key={ appointment .id }
                            appointment={ appointment }
                            deleteAppointmentAttended={ deleteAppointmentAttended }
                        />
                    ))}
                </div>
            </div>
        </div>
    </Fragment>
  );
}

export default App;
