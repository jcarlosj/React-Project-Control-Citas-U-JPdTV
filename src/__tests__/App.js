import React from 'react';

/** Testing Library Dependencies */
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom/extend-expect';       //  Para dar soporte a toBeInTheDocument()

/** Components */
import App from '../App';

describe( '<App />', () => {

    test( 'Títulos coincidentes, Título Dinámico \'No hay citas pendientes\' ', () => {
        render( <App /> );      //  Monta Componente

        /** Verificar que los titulos existan en en DOM */
        const titles = [ 'Administrar citas', 'Solicitud de cita', 'No hay citas pendientes' ];
        
        for ( const title of titles ) {
            expect( screen .getByText( title ) ) .toBeInTheDocument();
        }

        /** Verificar etiqueta y contenido del titulo principal de la aplicacion */
        const elTitle = screen .getByTestId( 'app-title' );

        expect( elTitle .tagName ) .toBe( 'H1' );               //  Verifica que la etiqueta sea H1
        expect( elTitle .textContent ) .toBe( titles[ 0 ] );    //  Verifica que el texto de en la etiqueta sea 'Administrar citas'

        /** Verifica Titulo Dinámico cuando no hay citas */
        expect( screen .getByTestId( 'app-dynamic-title' )  .textContent ) .not .toBe( 'Listado citas' );
        expect( screen .getByTestId( 'app-dynamic-title' )  .textContent ) .toBe( 'No hay citas pendientes' );

    });

    test( 'Cambia el título dinámico a \'Hay citas pendientes\'', () => {
        render( <App /> );      //  Monta Componente

        /** Llena el formulario */
        userEvent .type( screen .getByTestId( 'pet-name' ), 'Rex' );
        userEvent .type( screen .getByTestId( 'owner-name' ), 'Juan Carlos' );
        userEvent .type( screen .getByTestId( 'date' ), '2021-09-10' );
        userEvent .type( screen .getByTestId( 'hour' ), '09:10' );
        userEvent .type( screen .getByTestId( 'symptoms' ), 'Duerme como un angelito' );


        const btnSubmit = screen .getByTestId( 'btn-submit' );
        userEvent .click( btnSubmit );      //  Envia un evento al elemento (Hacemos clic - Forma Anterior

        /** Alert-Error */
        expect(
            screen .queryByTestId( 'alert-error' )        //  No encuentre el elemento. Se usa queryByTestId cuando no hay certeza de encontrar el elemento y evitar el error con getByTestId
        ) .not.toBeInTheDocument();                                             //  En este caso no se debe negar usando not, si no agregando 'does-not-exist' a queryByTestId

        /** Verificamos el titulo dinamico */
        expect( screen .getByTestId( 'app-dynamic-title' ) .textContent ) .toBe( 'Listado citas' );
        expect( screen .getByTestId( 'app-dynamic-title' ) .textContent ) .not .toBe( 'No hay citas pendientes' );
    });

    test( 'Verificar la inserción de una cita en el DOM', async () => {
        render( <App /> );      //  Monta Componente

        const appointments = await screen .findAllByTestId( 'app-apointment' );

        // console .log( appointments );   //  Promise { <pending> } si no se usa async/await

        //  Snapshop: Crea un archivo con la estructura del DOM para verificar su contenido
        // expect( appointments ) .toMatchSnapshot();      //   Asegura que un valor coincida con la instantánea más reciente. (ver nuevo directorio __snapshops__ )

        expect( screen .getByTestId( 'btn-atender' ) ) .toBeInTheDocument();        //  Verifica que existe el elemento
        expect( screen .getByTestId( 'btn-atender' ) .tagName ) .toBe( 'BUTTON' );  //  Verifica que el elemento es BUTTON

        // Verificar cita insertada
        expect( screen .getByText( 'Rex' ) ).toBeInTheDocument();
    });

});