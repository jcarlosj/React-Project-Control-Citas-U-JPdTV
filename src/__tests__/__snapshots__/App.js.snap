// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> Verificar la inserción de citas en el DOM 1`] = `
Array [
  <table
    class="u-full-width"
    data-testid="app-apointment"
  >
    <thead>
      <tr>
        <th>
          Mascota
        </th>
        <th>
          Propietario
        </th>
        <th>
          Fecha - Hora
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Rex
        </td>
        <td>
          Juan Carlos
        </td>
        <td>
          2021-09-10
           - 
          09:10
           
        </td>
      </tr>
      <tr>
        <th
          colspan="3"
        >
          Sintomatología
        </th>
      </tr>
      <tr>
        <td
          colspan="3"
        >
          Duerme como un angelito
        </td>
      </tr>
      <tr>
        <td
          colspan="3"
        >
          <button
            class="u-full-width button"
          >
            atender ×
          </button>
        </td>
      </tr>
    </tbody>
  </table>,
  <table
    class="u-full-width"
    data-testid="app-apointment"
  >
    <thead>
      <tr>
        <th>
          Mascota
        </th>
        <th>
          Propietario
        </th>
        <th>
          Fecha - Hora
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Jacobo
        </td>
        <td>
          Ana Maria
        </td>
        <td>
          2021-09-10
           - 
          10:10
           
        </td>
      </tr>
      <tr>
        <th
          colspan="3"
        >
          Sintomatología
        </th>
      </tr>
      <tr>
        <td
          colspan="3"
        >
          Juega y salta sin parar angelito
        </td>
      </tr>
      <tr>
        <td
          colspan="3"
        >
          <button
            class="u-full-width button"
          >
            atender ×
          </button>
        </td>
      </tr>
    </tbody>
  </table>,
]
`;
